### PR TITLE
Adding a new workload oltp_multi_value_insert which perform multiple rows insert with a single insert statement

### PR DESCRIPTION
**Pull request by shantanugupta-yb**
This workload only support prepare, bulk_load and cleanup command. 
Also in "bulk_load" phase it will only load data using single thread on a single table. I have added warning message for same

Below are the commands I have tested:
1. Command to perform cleanup
`sysbench oltp_multi_value_insert --pgsql-host=172.151.51.78 --table_size=12000 --serial_cache_size=1000 --range_selects=false --time=7200 --create_secondary=true --warmup-time=0 --tables=1 --range_key_partitioning=false --thread-init-timeout=3000 --pgsql-db=yugabyte --pgsql-user=yugabyte --db-driver=pgsql --pgsql-port=5433 --threads=1 --batch_insert_count=4000 cleanup`

2. Command to create tables
`sysbench oltp_multi_value_insert --pgsql-host=172.151.51.78 --table_size=12000 --serial_cache_size=1000 --range_selects=false --time=7200 --create_secondary=true --warmup-time=0 --tables=1 --range_key_partitioning=false --thread-init-timeout=3000 --pgsql-db=yugabyte --pgsql-user=yugabyte --db-driver=pgsql --pgsql-port=5433 --threads=1 --batch_insert_count=4000 prepare`

3. Command to run workload
`sysbench oltp_multi_value_insert --pgsql-host=172.151.51.78 --table_size=12000 --serial_cache_size=1000 --range_selects=false --time=7200 --create_secondary=true --warmup-time=0 --tables=1 --range_key_partitioning=false --thread-init-timeout=3000 --pgsql-db=yugabyte --pgsql-user=yugabyte --db-driver=pgsql --pgsql-port=5433 --threads=1 --batch_insert_count=4000 bulk_load`

> sysbench 1.1.0-c7af404 (using bundled LuaJIT 2.1.0-beta3)
> 
> Server Endpoint: 172.151.51.78 
> oltp_multi_value_insert creates a single table and bulk inserts are performed using single thread
> ( 7:44:10) Inserting 12000 records into 'sbtest1'
> Insert start from 1 to 12000
> ( 7:44:10) Done Loading

### CODE REVIEW

**Comment by hbhanawat** [comment]
@shantanugupta-yb I think conceptually the lua file that you created doesn't seem correct to me. For perf measurements, we have create, load and execute phase. Why do you want to do the measurements in the load phase and not in the execute phase? The code that you have written should go in the event() function of a new lua file. I created this lua file and it seems to be doing the expected:
```

#!/usr/bin/env sysbench

 
require("oltp_common")
cursize=0

function prepare_statements()
    if not sysbench.opt.skip_trx then
        prepare_begin()
        prepare_commit()
    end
    prepare_delete_inserts()
end

function event()
   local table_name = "sbtest" .. sysbench.rand.uniform(1, sysbench.opt.tables)

   iquery = ""

   if sysbench.opt.auto_inc then
      iquery = "INSERT INTO " .. table_name .. "(k, c, pad) VALUES"
   else
      iquery = "INSERT INTO " .. table_name .. "(id, k, c, pad) VALUES"
   end

   con:bulk_insert_init(iquery)

   local c_val
   local pad_val

   cursize = 1

   for i = 1, sysbench.opt.batch_insert_count do

       c_val = get_c_value()
       pad_val = get_pad_value()

       if (sysbench.opt.auto_inc) then
          query = string.format("(%d, '%s', '%s')",
                                sysbench.rand.default(1, sysbench.opt.table_size),
                                c_val, pad_val)
       else
         -- to do - please complete
       end

       con:bulk_insert_next(query)
   end
   con:bulk_insert_done()
   con:bulk_insert_init(iquery)
 end
 
 function thread_done()
    con:bulk_insert_done()
    con:disconnect()
 end
 
```
The only challenge that I found with this approach is that if you set batch_insert_count greater than 1000 than it does not create the right number of values. This is because of the hardcoding in this file:
https://github.com/yugabyte/sysbench/blob/c7af4040b72e0f6b5f01229e12e829c6895a4a32/src/db_driver.c#L46

However, your code will hit this challenge too. 

PS: please test the code entirely. I have tested only the main code path.

**Comment by shantanugupta-yb** [comment]
The whole purpose of custome LUA scripts is to enable user to write workload for a usecase which is not natively supported by sysbench. Also sysbench in itself does not imply any bindings on the phases that has to be part of a given workload. It seems logical to have these phases even for custom sysbench workloads so that we can leverage the current perf service framework without making any code changes to perf service.

I have reorganised my existing code such that it supports the same set of phases. Also confirmed that I am able to support 4k values in a single insert operation. 

```
[centos@ip-10-9-207-41 sysbench]$ sysbench oltp_multi_value_insert --pgsql-host=172.151.53.130 --table_size=100000 --serial_cache_size=1000 --range_selects=false --time=60 --create_secondary=true --warmup-time=0 --tables=1 --range_key_partitioning=false --thread-init-timeout=3000 --pgsql-db=yugabyte --pgsql-user=yugabyte --db-driver=pgsql --pgsql-port=5433 --threads=2 --batch_insert_count=4000 --skip_trx=on run
sysbench 1.1.0-c7af404 (using bundled LuaJIT 2.1.0-beta3)

Running the test with following options:
Number of threads: 2
Initializing random number generator from current time


Initializing worker threads...

Server Endpoint: 172.151.53.130 
Threads started!

SQL statistics:
    queries performed:
        read:                            0
        write:                           446
        other:                           0
        total:                           446
    transactions:                        446    (7.41 per sec.)
    queries:                             446    (7.41 per sec.)
    ignored errors:                      0      (0.00 per sec.)
    reconnects:                          0      (0.00 per sec.)

Throughput:
    events/s (eps):                      7.4080
    time elapsed:                        60.2048s
    total number of events:              446

Latency (ms):
         min:                                  230.75
         avg:                                  269.96
         max:                                  394.26
         95th percentile:                      308.84
         sum:                               120403.90

Threads fairness:
    events (avg/stddev):           223.0000/0.00
    execution time (avg/stddev):   60.2019/0.00 
```

File: src/lua/oltp_common.lua [hunk 1]
```diff
@@ -117,6 +118,14 @@ function cmd_load()
    end
 end
 
+function execute_multi_value_insert()
+   local drv = sysbench.sql.driver()
+   local con = drv:connect()
+--    sysbench.opt.tables = 1
```
Comment (hbhanawat) [inline src/lua/oltp_common.lua:124]: remove the commented code

File: src/lua/oltp_multi_value_insert.lua [hunk 2]
```diff
@@ -0,0 +1,33 @@
+#!/usr/bin/env sysbench
+
+-- ----------------------------------------------------------------------
+-- Multi-rows in a single Insert OLTP benchmark
+-- ----------------------------------------------------------------------
+
+require("oltp_common")
+
+sysbench.cmdline.commands.prepare = {
```
Comment (hbhanawat) [inline src/lua/oltp_multi_value_insert.lua:9]: Needed?

File: src/lua/oltp_multi_value_insert.lua [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+         sysbench.opt.table_size=0
+      end
+
+      cmd_create()
+      --cmd_prepare()
+   end,
+   sysbench.cmdline.PARALLEL_COMMAND
+}
+
+function prepare_statements()
```
Comment (hbhanawat) [inline src/lua/oltp_multi_value_insert.lua:24]: needed?

File: src/lua/oltp_common.lua [hunk 4]
```diff
@@ -270,6 +290,60 @@ CREATE TABLE sbtest%d(
...
+           c_val = get_c_value_smaller()
+           pad_val = get_pad_value_smaller()
+       end
+
+       if (sysbench.opt.auto_inc) then
+          query = string.format("(%d, '%s', '%s')",
+                                sysbench.rand.default(1, sysbench.opt.table_size),
+                                c_val, pad_val)
+       else
+          i = sysbench.rand.unique() - 2147483648
```
Comment (hbhanawat) [inline src/lua/oltp_common.lua:328]: will not work as you are modifying i

**Review by hbhanawat — CHANGES_REQUESTED** [review]

File: src/lua/oltp_common.lua (hunk 1 above)
Comment (shantanugupta-yb) [reply src/lua/oltp_common.lua]: Done

File: src/lua/oltp_multi_value_insert.lua (hunk 2 above)
Comment (shantanugupta-yb) [reply src/lua/oltp_multi_value_insert.lua]: Removed.

File: src/lua/oltp_multi_value_insert.lua (hunk 3 above)
Comment (shantanugupta-yb) [reply src/lua/oltp_multi_value_insert.lua:9]: Yes this is needed, oltp_common.sh expects this function to be defined.

File: src/lua/oltp_common.lua (hunk 4 above)
Comment (shantanugupta-yb) [reply src/lua/oltp_common.lua]: Scope of "i" dont impact the loop. Though I have changed the logic here. 

The logic of `sysbench.rand.unique() - 2147483648` does not work in this case since it results in conflict for few key which are part of 4k batch(or lesser) as in load phase we have already inserted 50M rows. 

With new logic, I increment the row count serially above the #rows inserted part of load phase.

File: src/lua/oltp_common.lua [hunk 5]
```diff
@@ -270,6 +292,60 @@ CREATE TABLE sbtest%d(
    end
 end
 
+function bulk_inserts(con, table_num)
+
+   if (sysbench.opt.table_size > 0) then
+      time = os.date("*t")
```
Comment (hbhanawat) [inline src/lua/oltp_common.lua:298]: Why is this needed?

File: src/lua/oltp_common.lua [hunk 6]
```diff
@@ -270,6 +292,60 @@ CREATE TABLE sbtest%d(
...
+       if (sysbench.opt.auto_inc) then
+          query = string.format("(%d, '%s', '%s')",
+                                sysbench.rand.default(1, sysbench.opt.table_size),
+                                c_val, pad_val)
+       else
+          query = string.format("(%d, %d, '%s', '%s')",
+                                sysbench.opt.table_size+pk_number,
+                                sysbench.rand.default(1, sysbench.opt.table_size),
+                                c_val, pad_val)
+          pk_number=pk_number+1
```
Comment (hbhanawat) [inline src/lua/oltp_common.lua:335]: This will not work in multi threaded scenarios.

Comment (hbhanawat) [reply src/lua/oltp_common.lua:335]: We need to have the code which we are using in bulk_load. What was the problem with the previous approach of using a number between and 1 and MAX_INT

File: src/lua/oltp_common.lua [hunk 7]
```diff
@@ -270,6 +292,60 @@ CREATE TABLE sbtest%d(
...
+       end
+
+       if (cursize % sysbench.opt.batch_insert_count ~= 0) then
```
Comment (hbhanawat) [inline src/lua/oltp_common.lua:338]: Why is this check needed? The "for" loop runs for batch_insert_count times. We can do con:bulk_insert_done() outside of for loop and con:bulk_insert_next(query) inside the for loop

File: src/lua/oltp_common.lua [hunk 8]
```diff
@@ -270,6 +292,60 @@ CREATE TABLE sbtest%d(
...
+          con:bulk_insert_next(query)
+       else
+          con:bulk_insert_next(query)
+          con:bulk_insert_done()
+       end
+
+       cursize = cursize + 1
```
Comment (hbhanawat) [inline src/lua/oltp_common.lua:345]: cursize is not needed as we are inside a for loop.

File: src/lua/oltp_common.lua [hunk 9]
```diff
@@ -117,6 +122,12 @@ function cmd_load()
    end
 end
 
+function execute_multi_value_insert()
+   local drv = sysbench.sql.driver()
```
Comment (hbhanawat) [inline src/lua/oltp_common.lua:126]: Why are you creating a connection every time execute_multi_value_insert is called. Connection should be created only once. It is already being created in the thread_init so you might have con available here already.

**Review by hbhanawat — CHANGES_REQUESTED** [review]

File: src/lua/oltp_common.lua (hunk 6 above)
Comment (hbhanawat) [reply src/lua/oltp_common.lua:335]: Just read your previous comment. In that case, we can't do bulk insert in its current mode for non auto_inc cases. I could recommend. Throwing an error for this case rather than supporting it partially.

File: src/lua/oltp_common.lua [hunk 10]
```diff
@@ -117,6 +122,12 @@ function cmd_load()
...
+   local con = drv:connect()
+   bulk_inserts(con, 1)
```
Comment (hbhanawat) [inline src/lua/oltp_common.lua:128]: You should be picking the table number randomly everytime.

File: src/lua/oltp_multi_value_insert.lua [hunk 11]
```diff
@@ -0,0 +1,31 @@
...
+function thread_init()
+    if (not sysbench.opt.auto_inc ) then
+       error("Workload oltp_multi_value_insert only support auto-increment of Primary Key column i.e. (auto_inc=true)")
+    end
+    drv = sysbench.sql.driver()
+    con = drv:connect()
+end
+
+function event()
+    if (sysbench.opt.auto_inc) then
```
Comment (hbhanawat) [inline src/lua/oltp_multi_value_insert.lua:19]: this check is not needed here as we are anyway checking it above

**Review by hbhanawat — APPROVED** [review]
Approving as the comments are mostly formatting related. Please fix the comments before you merge this PR.
